pkg/server/commands/listusers/list_users.go

```diff
@@ -12,3 +14,87 @@ type listUsersRequest interface {
 	GetUserFilters() []*openfgav1.ListUsersFilter
 	GetContextualTuples() *openfgav1.ContextualTupleKeys
 }
+
+type internalListUsersRequest struct {
+	StoreID              string
+	AuthorizationModelID string
+	Object               *openfgav1.Object
+	Relation             string
+	UserFilters          []*openfgav1.ListUsersFilter
+
+	ContextualTuples *openfgav1.ContextualTupleKeys
+
+	// visitedUsersetsMap keeps track of the "path" we've made so far.
+	// It prevents stack overflows by preventing visiting the same userset twice.
+	visitedUsersetsMap map[string]struct{}
```

Comment:
**Question:** Am I to understand that cycles are _only_ caused by usersets? And if so, can you explain why?

Reply:
No, usersets don't _cause_ the cycles, they are merely the way of detecting if we've encountered one.

Some cycles are caused by the model, doesn't matter which tuples you have (in red below), e.g.

```
define a: b
define b: a
```

If the request is ListUsers(document:1 , a, filter = [user]), we start expanding `document:1#a`, so we mark it as visited, then we go to `document:1#b`, we mark it as visited, then we will go back to `document:1#a`, loop!


![image](https://github.com/openfga/openfga/assets/5374887/5f2585bd-ad42-4544-8f62-a6b5699d12f5)

The other category of cycles are caused by the model _and_ the tuples in the store (in yellow above).

If the request is ListUsers(document:1 , viewer, filter = [user]), we start expanding `document:1#viewer`, so we mark it as visited, then by virtue of the computed userset we go to `document:1#editor`, we mark it as visited, then assume we have a tuple `document:1#viewer@document:1#editor`, we will go back to `document:1#viewer`, loop!

Reply:
I'm getting confused by usage of usersets in `visitedUsersetsMap`. The first example doesn't have user sets. Can we keep consistent with check's `VisitedPaths`?

Reply:
> The first example doesn't have user sets

Yes and no... there are no tuples, true, but with the query being `document:1 , a, filter = [user]`, I need to expand the userset `document:1#a`.

> Can we keep consistent with check's VisitedPaths?

No, Check uses full tuples as key of the map, but I don't have that here. I am using the same key as in ListObjects: 

https://github.com/openfga/openfga/blob/0efd7244c43197369337ce288d6af6b6245f9fe5/pkg/server/commands/reverseexpand/reverse_expand.go#L271-L275